main/journal_to_datastore_and_journal.es6.js

```diff
@@ -0,0 +1,233 @@
+      pkg.ReleaseWebInterfaceJunction.id, pkg.ReleaseWebInterfaceJunction, ctx),
+  browserMetricsDAO: getReadJournalDAO(
+      pkg.BrowserMetricData.id, pkg.BrowserMetricData, ctx),
+  apiVelocityDAO: getReadJournalDAO(
+      pkg.ApiVelocityData.id, pkg.ApiVelocityData, ctx),
+}, ctx);
+
+
+// Context for reading from / writing to Datastore.
+//
+// TODO(markdittmer): Figure out why partitioning isn't working. This component
+// was originally to be instantiated in context:
+//
+// foam.__context__.createSubContext({
+//   datastoreNamespaceId: 'someDatastoreNamespace'
+// })
+//
+// but it was discovered that put()s were going to the default namespace, even
+// though queries were correctly scoped to 'someDatastoreNamespace'.
+const datastoreCtx = pkg.DatastoreContainer.create({
+  mode: pkg.DatastoreContainerMode.DATA_COLLECTOR,
+  gcloudAuthEmail: credentials.client_email,
+  gcloudAuthPrivateKey: credentials.private_key,
+  gcloudProjectId: credentials.project_id,
+  logger: logger,
+}).ctx;
+
+// Replace Datastore context's SyncDAO.delegate values with JDAOs to retain
+// versioned copy of latest data.
+datastoreCtx.releaseDAO.delegate = getOverwriteJournalDAO(
+    pkg.VersionedRelease.id, pkg.VersionedRelease, ctx);
+datastoreCtx.webInterfaceDAO.delegate = getOverwriteJournalDAO(
+    pkg.VersionedWebInterface.id, pkg.VersionedWebInterface, ctx);
+datastoreCtx.releaseWebInterfaceJunctionDAO.delegate = getOverwriteJournalDAO(
+    pkg.VersionedReleaseWebInterfaceJunction.id,
+    pkg.VersionedReleaseWebInterfaceJunction, ctx);
+datastoreCtx.browserMetricsDAO.delegate = getOverwriteJournalDAO(
+    pkg.VersionedBrowserMetricData.id, pkg.VersionedBrowserMetricData, ctx);
+datastoreCtx.apiVelocityDAO.delegate = getOverwriteJournalDAO(
+    pkg.VersionedApiVelocityData.id, pkg.VersionedApiVelocityData, ctx);
+
+// SyncDAOs connected to Datastore.
+const releaseSyncDAO = datastoreCtx.releaseDAO;
+const webInterfaceSyncDAO = datastoreCtx.webInterfaceDAO;
+const releaseWebInterfaceJunctionSyncDAO =
+    datastoreCtx.releaseWebInterfaceJunctionDAO;
+const browserMetricsSyncDAO = datastoreCtx.browserMetricsDAO;
+const apiVelocitySyncDAO = datastoreCtx.apiVelocityDAO;
+
+// An unversioned cache of current Datastore.
+const releaseCacheDAO = cacheCtx.releaseDAO;
+const webInterfaceCacheDAO = cacheCtx.webInterfaceDAO;
+const releaseWebInterfaceJunctionCacheDAO =
+    cacheCtx.releaseWebInterfaceJunctionDAO;
+const browserMetricsCacheDAO = cacheCtx.browserMetricsDAO;
+const apiVelocityCacheDAO = cacheCtx.apiVelocityDAO;
+
+// In-memory DAOs of new data to be imported.
+const releaseImportDAO = importCtx.releaseDAO;
+const webInterfaceImportDAO = importCtx.webInterfaceDAO;
+const releaseWebInterfaceJunctionImportDAO =
+    importCtx.releaseWebInterfaceJunctionDAO;
+const browserMetricsImportDAO = importCtx.browserMetricsDAO;
+const apiVelocityImportDAO = importCtx.apiVelocityDAO;
+
+//
+// Generic algorithm for data import:
```

Comment:
This sounds less than trivial. Can't we write all the data to a new thing, and then point the old thing at the new thing? (Yay for comments, without it I probably wouldn't have known what I was looking at.)

Reply:
Datastore generally uses partitions for that. There's an outstanding bug to investigate why FOAM's Datastore/REST implementation does not handle partitions properly. It's a particularly frustrating problem because our tests against Datastore's emulator suggests that partitions ARE working properly, but the don't work in production.

Reply:
OK, sounds like that is a bigger project in any event, not today.